### PR TITLE
Pauls/add fargate task support

### PR DESCRIPTION
**Pull request by PaulSkersis**

### CODE REVIEW

File: ecs.tf [hunk 1]
```diff
@@ -5,12 +5,13 @@ data "aws_ecs_cluster" "target_cluster" {
 resource "aws_ecs_task_definition" "main_task" {
   family                   = "${var.app_name}-tsk"
   requires_compatibilities = [var.launch_type]
-  network_mode             = "bridge"
+  network_mode             = var.network_mode
```
Comment (dumptruckman) [inline ecs.tf]: If this has to be set to a certain value for fargate to work we could make this a bit more automatic. You could put a conditional here for the launch type and just set it to `awsvpc` if we're doing fargate.

**Review by kcchoate — DISMISSED** [review]

File: variables.tf [hunk 2]
```diff
@@ -201,3 +213,14 @@ variable "circuit_breaker_sns_topic_arn" {
   default     = null
   description = "The arn of the SNS topic to publish deployment circuit breaker failure messages to. If not provided, a SNS topic will be provided by this module."
 }
+
+variable "network_configurations" {
+  type        = list(object({ subnets = list(string), security_groups = list(string), assign_public_ip = bool }))
+  description = "A list of definitions to attach network configurations to the ECS task."
+}
```
Comment (dumptruckman) [inline variables.tf]: I think it would be easier to read the type here if there were some line breaks.

Comment (dumptruckman) [reply variables.tf]: Also, does this variable break things without a value?

File: lb.tf [hunk 3]
```diff
@@ -67,7 +67,7 @@ resource "aws_lb_target_group" "lb_targets" {
   port                 = 80
   protocol             = "HTTP"
   vpc_id               = var.vpc_id
-  target_type          = "instance"
+  target_type          = var.lb_target_type
```
Comment (dumptruckman) [inline lb.tf]: Once again, could automate this part if it has to be a specific value for fargate.

File: variables.tf (hunk 2 above)
Comment (PaulSkersis) [reply variables.tf]: Yeah fair, I copied this over from our Fargate singleton module: https://github.com/7Factor/terraform-fargate-ecs-singleton-task/blob/master/variables.tf 
Let me see how it interacts if there's no value provided. I think if I give it a default of an empty list it should be fine.

File: lb.tf (hunk 3 above)
Comment (PaulSkersis) [reply lb.tf]: Updated

File: ecs.tf (hunk 1 above)
Comment (PaulSkersis) [reply ecs.tf]: Updated

File: ecs.tf [hunk 4]
```diff
@@ -5,12 +5,13 @@ data "aws_ecs_cluster" "target_cluster" {
 resource "aws_ecs_task_definition" "main_task" {
   family                   = "${var.app_name}-tsk"
   requires_compatibilities = [var.launch_type]
-  network_mode             = "bridge"
+  network_mode             = var.launch_type == "FARGATE " ? "awsvpc" : var.network_mode
```
Comment (dumptruckman) [inline ecs.tf:8]: If the launch type isn't required to be all caps for fargate then you might want to consider something like this:
```suggestion
  network_mode             = lower(var.launch_type) == "fargate " ? "awsvpc" : var.network_mode
```

Comment (dumptruckman) [reply ecs.tf:8]: Though if it has to be all caps this is fine already ofc.

Comment (PaulSkersis) [reply ecs.tf:8]: Pretty sure it's required to be all caps. All the docs reference it as such, although admittedly I haven't tried it in lower.

File: variables.tf (hunk 2 above)
Comment (dumptruckman) [reply variables.tf]:
```suggestion
variable "network_configurations" {
  type = list(object({
    subnets          = list(string),
    security_groups  = list(string),
    assign_public_ip = bool
  }))
  description = "A list of definitions to attach network configurations to the ECS task."
}

```

File: ecs.tf (hunk 4 above)
Comment (dumptruckman) [reply ecs.tf:8]: Alternately, I guess you could use `upper` just to be safe.

Comment (kcchoate) [reply ecs.tf:8]: Is it possible to only allow specific values for inputs to modules? You could limit the launch type to be exactly `FARGATE` or `EC2` if you can, which removes this as a problem.

Comment (PaulSkersis) [reply ecs.tf:8]: https://stackoverflow.com/questions/57190035/terraform-how-to-restrict-an-input-variable-to-a-list-of-possible-choices
Looks like we can, and give a custom error message too which is pretty neat.

Comment (kcchoate) [reply ecs.tf:8]: Looks like the closest you can get is [Custom Validation Rules](https://developer.hashicorp.com/terraform/language/values/variables)

File: variables.tf (hunk 2 above)
Comment (PaulSkersis) [reply variables.tf]: Updated to include a default value of an empty list and prettified. Might be worth going through after this PR in a follow-up to fix formatting, every pre-existing list is smushed onto one line.

File: ecs.tf (hunk 4 above)
Comment (PaulSkersis) [reply ecs.tf:8]: Added validation for launch type

File: variables.tf [hunk 5]
```diff
@@ -153,6 +160,10 @@ variable "service_deployment_minimum_healthy_percent" {
 variable "launch_type" {
   default     = "EC2"
   description = "The launch type for the task. We assume EC2 by default."
+  validation {
+    condition     = contains(["EC2", "FARGATE"], var.launch_type)
+    error_message = "Only EC2 and FARGATE are supported as launch types by this module."
+  }
```
Comment (dumptruckman) [inline variables.tf:166]: Nice. This was definitely a good thought.

**Review by dumptruckman — APPROVED** [review]

**Review by kcchoate — APPROVED** [review]

**Pull request merged**